### DISCUSSION TITLE
아래 코드에서 출력이 어떻게되는 지와 이 코드의 문제점을 설명하세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#88] [클로저]
- 아래 코드에서 출력이 어떻게되는 지와 이 코드의 문제점을 설명하세요.
  ```js
   const Person = (function () {
    let _age = 0;
    
    function Person(name, age) {
      this.name = name;
      _age = age;
    }
  
    Person.prototype.sayHi = function () {
      console.log(`Hi! My name is ${this.name}. I am ${_age}.`);
    };
  
    return Person;
  }());
  
  const me = new Person('Lee', 20);
  me.sayHi();
  
  const you = new Person('Kim', 30);
  you.sayHi();
  
  me.sayHi();
  ```

**Comment by pipisebastian** [comment]
> * 아래 코드에서 출력이 어떻게되는 지와 이 코드의 문제점을 설명하세요.
>   ```js
>    const Person = (function () {
>     let _age = 0;
>     
>     function Person(name, age) {
>       this.name = name;
>       _age = age;
>     }
>   
>     Person.prototype.sayHi = function () {
>       console.log(`Hi! My name is ${this.name}. I am ${_age}.`);
>     };
>   
>     return Person;
>   }());
>   
>   const me = new Person('Lee', 20);
>   me.sayHi();
>   
>   const you = new Person('Kim', 30);
>   you.sayHi();
>   
>   me.sayHi();
>   ```

- Hi! My name is Lee. I am 20.
- Hi! My name is Kim. I am 30.
- Hi! My name is Lee. I am 30. 
-  `_age` 변수는 인스턴스마다 존재하는게 아닌, 즉시실행함수의 지역변수였고, 이는 실행이 끝나기에 클로저로써 동작합니다. 즉, 모든 인스턴스는 하나의 변수를 공유하고, 30이라는 값이 덮어씌어지게 됩니다. 
- + Person.prototype.sayHi 메서드는 프로토타입 메서드라서, 여러 객체가 이 메서드를 공유해서 씁니다. 그렇기에 me, you 인스턴스 둘다 sayHi 메서드를 공유하는데, 이때, _age를 참조할때, 덮어씌워진 _age를 가져오기 때문에, 30이 출력됩니다.

**Comment by publdaze** [comment]
> * 아래 코드에서 출력이 어떻게되는 지와 이 코드의 문제점을 설명하세요.
>   ```js
>    const Person = (function () {
>     let _age = 0;
>     
>     function Person(name, age) {
>       this.name = name;
>       _age = age;
>     }
>   
>     Person.prototype.sayHi = function () {
>       console.log(`Hi! My name is ${this.name}. I am ${_age}.`);
>     };
>   
>     return Person;
>   }());
>   
>   const me = new Person('Lee', 20);
>   me.sayHi();
>   
>   const you = new Person('Kim', 30);
>   you.sayHi();
>   
>   me.sayHi();
>   ```
### 생성자 함수
- Person을 살펴보면 즉시 실행 함수 내부에 중첩함수의 Person이 존재하는데,
- 이 Person에는 외부함수 즉, 즉시 실행 함수의 지역변수인 _age를 참조하고 있으며
- 이 Person 함수가 리턴되므로 클로저라고 볼 수 있음
### me 인스턴스
- me라는 변수에 생성자 함수의 인스턴스가 할당되는데,
- 해당 인스턴스는 생성자 함수의 this 바인딩이 인스턴스 객체를 참조하므로, 인스턴스 독자적인 this의 프로퍼티, name에 'Lee'가 할당된 값을 가지고
- 생성자 함수의 두 번째 인수로 들어간 20은 상위 렉시컬 환경에서 _age 식별자가 검색되어 할당됨
- 이후 me.sayHi를 호출하고 this.name은 인스턴스에 바인딩된 this로 인해 'Lee'로 평가되고
- sayHi 메서드도 클로저이므로 이미 생명주기가 끝난 외부 함수에 대한 식별자 참조가 가능하여 _age는 20으로 평가됨
- 따라서 출력 결과는 `Hi! My name is Lee. I am 20.`
### you 인스턴스
- 다음 you 변수에 'Kim', 30을 인수로 갖는 생성자 함수의 인스턴스가 할당되는데,
- 해당 인스턴스는 생성자 함수는 인스턴스 독자적인 this의 프로퍼티 name에 'Kim'이 할당되고,
- 인스턴스가 아닌 생성자 함수의 _age 식별자에 30이 할당됨
- 이후 you.sayHi를 호출하고 this.name은 인스턴스에 바인딩된 this로 인해 'Kim'으로 평가되고
- _age는 30으로 평가되어
- 출력 결과는 `Hi! My name is Kim. I am 30.`
### me.sayHi 재호출
- 생성자 함수의 prototype 메서드로 인스턴스 메소드와 달리 함수 실행 시 마다 생성되는 게 아닌 해당 생성자의 모든 인스턴스가 같은 prototype 메서드를 공유함
- 따라서 해당 메서드가 평가될 때 this.name은 독자적인 this 바인딩으로 'Lee'가 되지만, _age는 바로 이전에 평가되었던 30이 되어
- `Hi! My name is Lee. I am 30.`이 출력됨
### 문제점
- 원래 기대하던 결과는 `Hi! My name is Lee. I am 20.`으로 인스턴스 별로 독자적인 값을 가지길 원하는데 의도하지 않은 결과가 나온다는 문제점이 있음
